### ISSUE TITLE
Design & implement slashing rules for missing pre-images

### ISSUE THREAD

**Issue by Geod24** [bosagora/agora#1076]
Currently, we are building rules based on the availability of pre-images: signature scheme and quorum balancing.
We also have a timer in place to ensure our pre-images are available long before we need them.
Now we need to add a way to detect and log whenever pre-images are missing, decide on a policy of what to do with it, and implement it.

### Detection

The first step IMO is to detect whenever we are missing a pre-image. This will allow us to make sure we have everything properly implemented at the moment. Once this is the case, we can move on to writing a test case.
That test case should have at least one node in the network doing the following:
- (A) Never send any pre-image after its initial enrollment;
- (B) Only send one pre-image after enrolling then no more;
- (C) Send pre-image for half of its cycle (potentially in one burst) then stop sending pre-image;
- (D) Send pre-image for half its cycle, stop it for a quarter of the cycle, resume until the end;
Those 4 situations should be tested *independently*, so not during the same cycle, and not using the same utxo key / public key.
This will allow us to see how the network currently behaves when faced with those situations, and correct any remaining issue.

### Policy decision

Once we have that basis in place, we will need to come up with the slashing policy for missing pre-images.
As discussed in the previous meetings, missing pre-images is one of the gravest mistakes a validator can commit, short of outright hostile actions (such as signing two blocks for the same height).

This will need to be further discussed with the team, but currently my mind is on two things:
- gradual punishment for validators missing to relay their pre-image, so that they still have an incentive to come back and propagate them if they miss a deadline. The more pre-images, the safer the randomness function.
- applying a loss-of-income penalty to the quorum: this gives a strong incentive to the quorum to *keep and propagate* pre-images for nodes in their quorum, ensuring that they will not simply ignore them to maximize the penalty of a validator;

### Implementation

Once we have the basic tests cases, and decided on a policy, we  can move to an implementation. I expect the policy decision process to yield more test cases which we can implement.

**Comment by AndrejMitrovic** [comment]
> The more pre-images, the safer the randomness function.

One thing I still don't follow w.r.t. randomness: If all the preimages are revealed in advance then anyone can calculate the random value for any future block height (unless a new enrollment was added in the meantime).

How do we account for that? Is there another step that's missing?

**Comment by Geod24** [comment]
> If all the preimages are revealed in advance

What do you mean "are revealed in advance" ? There is currently no rule about forcing validator not to reveal all of them in advance, so yes validators could potentially reveal all 1008 pre-image when they enroll, but it only takes one honest validator to prevent anyone to calculate that random value.

Now why would we have one single honest validator ? I don't think that's too much of a requirement to begin with, but regardless, removing the randomness part means exposing yourself to attacks, because an attacker that is able to compute the random value for any block in your cycle will be able to position itself in the same quorum as you, potentially feeding you erroneous / bogus data, which you really don't want.

**Comment by AndrejMitrovic** [comment]
> but it only takes one honest validator to prevent anyone to calculate that random value.

I don't follow, how would they prevent it?

**Comment by AndrejMitrovic** [comment]
By only sending a few preimages in advance? But then they are the ones at risk of being slashed if they go offline.

**Comment by Geod24** [comment]
> By only sending a few preimages in advance? But then they are the ones at risk of being slashed if they go offline.

Yep. And yep.

**Comment by AndrejMitrovic** [comment]
What you're saying does however make sense. I didn't realize it until now, but we only need **1** good validator to enforce true randomness. That's neat.

**Comment by linked0** [comment]
> The first step IMO is to detect whenever we are missing a pre-image. This will allow us to make sure we have everything properly implemented at the moment. Once this is the case, we can move on to writing a test case.

The following test cases are being written now as the start of this issue.
(A) Never send any pre-image after its initial enrollment;
(B) Only send one pre-image after enrolling then no more;
(C) Send pre-image for half of its cycle (potentially in one burst) then stop sending pre-image;
(D) Send pre-image for half its cycle, stop it for a quarter of the cycle, resume until the end;

**Comment by linked0** [comment]
I need to focus on issue #849. So, I don't think I can spend this issue anymore in this sprint. Sorry for the delay.

**Comment by linked0** [comment]
The questions are answered by me. @Geod24 Would you please check it?

**What object(I mean `class`) slashes a bad validator?**
It can be all of the related objects like `Node`, `Ledger`, or `EnrollmentManager`.
**What is the penalty? What kind of loss does the validator receive?** 
We can send the deposit of 40K BOA to some defined address by publishing a new transaction.
**What is the loss-of-income penalty to the quorum?**
_Please explain it._

We can start with the first punishment option. After implementing the option, we can change the policy because I make the code more changeable for a new policy.

**Comment by AndrejMitrovic** [comment]
> It can be all of the related objects like Node, Ledger, or EnrollmentManager.

Please move slashing into a separate module. There are too many entangled parts of code living in Ledger and EnrollmentManager. It's going to be very hard to reason about code if we have complex monolith classes which have too many roles and responsibilities.

**Comment by linked0** [comment]
> Please move slashing into a separate module. There are too many entangled parts of code living in Ledger and EnrollmentManager. It's going to be very hard to reason about code if we have complex monolith classes which have too many roles and responsibilities.

Right. This should be a separate module.

**Comment by Geod24** [comment]
> What object(I mean class) slashes a bad validator?

As mentioned, a separate class.

> What is the penalty? What kind of loss does the validator receive?

After discussion, let's start with 10k on the missing pre-image, effective immediately.

> What is the loss-of-income penalty to the quorum?

This is slightly different: One way to punish validators is to slash (take away their stake), while the other way is not to remunerate them (not give them confirmation reward).

**Comment by linked0** [comment]
In a slashing situation, some transactions can send 10K of a frozen UTXO to the `Commons Budget' address and refund the remain amount for the validator. So I'm trying to add the information into a block about slashed validators, which is used to check the validity of transactions in new blocks. 
What do you think about adding a 'slashed_enrollments' field to a block header?

**Comment by Geod24** [comment]
> What do you think about adding a 'slashed_enrollments' field to a block header?

I was hoping to avoid this. Since the pre-image will be missing, we could use this as a sign that the transaction is valid.

**Comment by linked0** [comment]
> I was hoping to avoid this. Since the pre-image will be missing, we could use this as a sign that the transaction is valid.

But without the information, we need some operation determining whether a UTXO is the key of enrollments for every input in a proposed block before checking the pre-images are missing.
And there is another issue that occurs in the catch-up process. If we need to know of slashed validators, there's no way of identifying it in the catch-up process.
What do you think of it?

**Comment by Geod24** [comment]
> But without the information, we need some operation determining whether a UTXO is the key of enrollments for every input in a proposed block before checking the pre-images are missing.

When we look up the UTXO, we already know if the UTXO is frozen or not. So currently, such a transaction would be rejected.
If we make it valid, the check will rarely be performed.

> And there is another issue that occurs in the catch-up process. If we need to know of slashed validators, there's no way of identifying it in the catch-up process.

Do  we need to know of the slashed validators though, or rather, we need to know about a certain block being the "tip" ?

**Comment by linked0** [comment]
> When we look up the UTXO, we already know if the UTXO is frozen or not. So currently, such a transaction would be rejected.
> If we make it valid, the check will rarely be performed.

Right. We can use the information that the UTXO is frozen or not. The purpose of the checking code is to find out whether some transactions send 10K of a good validator's fund to `CommonsBudget` address intentionally, which is an invalid transaction. So we need to check that a pre-image for some validator is really missing on the consensus protocol.

> Do we need to know of the slashed validators though, or rather, we need to know about a certain block being the "tip"?

If we have a policy that slashed validators should be kept from being a validator again for some period like 1008 block, I thought It could be needed. But it's not an urgent feature we have to consider now.

**Comment by Geod24** [comment]
> If we have a policy that slashed validators should be kept from being a validator again for some period like 1008 block, I thought It could be needed. But it's not an urgent feature we have to consider now.

No  we  don't,  and we can't reliably implement it. They could have another stash of 40k around and use that. If they don't, then they have to wait for the 30k to be usable anyway.

**Comment by linked0** [comment]
Questions about implementing slashing rules.
1. Don't we need to sign the slashing transaction, which contains output for Commons Budget and one for the refund? Actually, there is not a proper key for signing it.
2. In the process of verifying transactions, the only criteria to check the validity of the slashing transaction is to check that the address of the first `output` is the CommonsBudget address. Of course, we should check the input of the transaction is frozen. And we also need to provide `isInvalidReason` function(agora.consensus.validation.Transaction) with the delegate for searching the candidates to be slashed because the validation function checks that the validator being slashed is the right candidate.

Please let me know your opinions.

**Comment by AndrejMitrovic** [comment]
> Don't we need to sign the slashing transaction, which contains output for Commons Budget and one for the refund? Actually, there is not a proper key for signing it.

In my opinion I don't think slashing should be a transaction. Because if it is, who is in charge of making the slashing transaction? You could say anyone, but now you'll have competing slashing transactions in the pool. Most importantly, what if nodes collude and refuse to add this transaction in the next block? Then the validator would not be slashed. 

Couldn't we make slashing a protocol-level rule? If block N is externalized and there's a missing preimage then automatically spend part of that node's frozen UTXO. 

---

There's a missing part here: we need proof that a preimage doesn't exist. So preimages would likely need to be stored in the block. There can be ways to optimize this with regards to storage space. For example, the `getBlocksFrom` API can blank out the preimages field in the blocks its returning and instead another API could return the last preimage for every cycle for the requested list of nodes. That way the receiver of the blocks can reconstruct an older preimage (and yes there would be issues with gaps, but it's not unsolvable). There are other proposals how to optimize this, but I think we can solve this later as a separate optimization issue?

**Comment by linked0** [comment]
> In my opinion I don't think slashing should be a transaction.

As @Geod24  and I discussed, we decided to make transactions for that. But there's some issue who is in charge of making the slashing transaction as you said. Currently, anyone who detects the missing pre-images can make transactions but it can be discussed more after my draft PR is submitted.

> Most importantly, what if nodes collude and refuse to add this transaction in the next block? Then the validator would not be slashed.

I thought the slashing transactions can be part of consensus because a proposed block could contain the transactions.

> Couldn't we make slashing a protocol-level rule? If block N is externalized and there's a missing preimage then automatically spend part of that node's frozen UTXO.

I will think about the protocol-level rule for slashing.

> There's a missing part here: we need proof that a preimage doesn't exist. So preimages would likely need to be stored in the block. 

I didn't think we could store pre-images into a block.

**Comment by AndrejMitrovic** [comment]
> I didn't think we could store pre-images into a block.

Then I don't understand how one can prove that a validator hasn't released their preimage.

**Comment by AndrejMitrovic** [comment]
> As @Geod24 and I discussed, we decided to make transactions for that. But there's some issue who is in charge of making the slashing transaction as you said. 

Well ok, if you discussed it and that's the best way forward, go for it. But it seems the discussions haven't really come up with a solid plan? There's still unanswered questions like the one's you've asked.

**Comment by linked0** [comment]
> > I didn't think we could store pre-images into a block.
> 
> Then I don't understand how one can prove that a validator hasn't released their preimage.

I think we could check the situation with the `getPreimageAt` of `ValidatorSet` while checking the validity of the slashing transactions in `isInvalidReason` of `agora.consensus.validation.Transaction` module. The function can be called with some delegate, of course.

**Comment by AndrejMitrovic** [comment]
We had a whiteboarding about this. The way the solution was proposed didn't make that much sense to me. There's a few issues:

- Signature verification of the slashing transaction becomes impossible / complex. Let's say the node which (allegedly) didn't send its preimage is "victim node" and the node which slashes is called "slashing node". The input UTXO belongs to the victim node, but the slashing node signs the slashing transaction. However now signature validation will fail. There is no information in the transaction itself about who actually signed the transaction, it is always assumed that the `Output` contains a signature by the key found in the matching `Input`. You would really need to have some kind of special support for slashing transactions where it can still be validated.

- If there is any kind of monetary incentive to include slashing transactions during nomination, then it's in each validator's self-interest to accept all slashing transactions of other nodes, even if they're fake. There is no punishment for sending fake slashing transactions (claims that preimages are missing even when they're not), yet there may be rewards if those slashing transactions are accepted.

**Comment by linked0** [comment]
> If block N is externalized and there's a missing preimage then automatically spend part of that node's frozen UTXO.

If we "automatically spend part of that node's frozen UTXO" after determining some pre-images are missing, doesn't it need to be a transaction? I don't see how the frozen UTXO can be spent without a transaction. Could you explain more about it?

**Comment by AndrejMitrovic** [comment]
> I don't see how the frozen UTXO can be spent without a transaction. Could you explain more about it?

The rules for what happens when a block is externalized are here:

https://github.com/bpfkorea/agora/blob/3a709988ac301333fbcc34c8eb5ac8d4f0b67eb9/source/agora/node/Ledger.d#L305

So for example, we currently update the UTXO set and the validator set:

https://github.com/bpfkorea/agora/blob/3a709988ac301333fbcc34c8eb5ac8d4f0b67eb9/source/agora/node/Ledger.d#L316-L317

There is nothing preventing us from adding a new rule where we check if a preimage is missing and we change the UTXO set again so part of the frozen funds are re-routed to the commons budget. It would just change the existing UTXO set.

**Comment by linked0** [comment]
@AndrejMitrovic  & I discussed the issues of the current approach last Friday.

In addition to [the result of our whiteboard](https://github.com/bpfkorea/agora/issues/1076#issuecomment-726572380), Another problem is how to confirm that a validator really didn't publish pre-images. In order to validate a slashing transaction in the validation module like `agora.consensus.validation.Transaction`, we can not help referring to `ValidatorSet` which is a very complex relationship because `ValidatorSet` also references the validation module.

And, when the catch-up process is running, there are not pre-images yet in the `ValidatorSet`, we can not use the validation module in that situation. 

So we could need to store pre-images in blocks as @AndrejMitrovic suggested. Any opinions welcome!

**Comment by AndrejMitrovic** [comment]
By the way, when I say "in the block", I don't necessarily mean physically putting all the preimages in the block. They may be stored in a separate data structure. But some minimal representation could be stored, for example a merkle root which would be part of the block hash.

I seem to recall that the next version of Ethereum has some idea of using preimages too. It would be good to take a look what their plan is.

**Comment by linked0** [comment]
I think saving the pre-image information in the block would the right approach for solving the current issues described above. So, I thought about how to save the pre-image of all validators. Here is my idea.

* We save pre-images into the next block which are only newly published in the current height.
* The structure of the information has the index of a validator, a 16-byte hash of the pre-image, and a distance of the pre-image
  * index of a validator: an index on an ordered list of current validators
  * hash of a pre-image: a pre-image can be shortened with one more hashing resulting in a 16-byte hash
  * distance of a pre-image: this is can be used for proofing the pre-image

It's different from @AndrejMitrovic 's but we can have a whiteboard session about it.

**Comment by bpalaggi** [comment]
@Geod24 Can we have your input on this?

**Comment by hewison-chris** [comment]
For the **Schnorr multisig** to work we also need access to the old `pre-images` when we are catching up (if we are to validate the block signature). 
I am thinking that maybe we create a new `preimage.dat` file to store the `pre-images` from each active validator. We don't need to store all the pre-images as we can hash the required number of times to get to older `pre-images`. We just keep updating when we receive new ones. When we restart and need to know the `pre-image`s we can request via a new client API the `pre-images` from other validators (who can fetch the info from their `preimage.dat` file. 

However we still need to know which `pre-images` were not revealed in time for slashing. This could be achieved by storing the info of missing validators not revealing `pre-images` in the block. This way the block will record the fact that enough of the validators agreed that the pre-image was missing at that point in time.

**Comment by linked0** [comment]
> I am thinking that maybe we create a new preimage.dat file to store the pre-images from each active validator.

All lastest pre-images for each validator are stored in the `ValidatorSet`.

> When we restart and need to know the pre-images we can request via a new client API the pre-images from other validators (who can fetch the info from their preimage.dat file.

The new API could be needed.

> This could be achieved by storing the info of missing validators not revealing pre-images in the block. 

That was my original idea. But there's some problem in reaching a consensus about who should be slashed because the delay of receiving pre-images could happen due to some network issues. And [@AndrejMitrovic suggested another issue about that](https://github.com/bpfkorea/agora/issues/1076#issuecomment-726572380). So we're thinking about saving pre-images into the block. @Geod24 will post his opinion on this.

**Comment by linked0** [comment]
@Geod24 Please leave your opinion when you have time.

**Comment by AndrejMitrovic** [comment]
I have a better idea, we'll have a whiteboarding session about this today and later I will draw some schematics in this issue.

**Comment by AndrejMitrovic** [comment]
## Synopsis

We have several different requirements in our system:

- We want to incentivize revealing preimages. The preimages are the foundation for the randomness in the system, which is needed for the quorum generator and will be used by other parts of the system to work correctly.
- We want to penalize not revealing preimages.
- We want proof that a preimage was indeed revealed for a specific block height. We do not want a situation where a preimage is revealed "after the fact" (after externalize). 

I propose an alternative to the original transaction-based preimage slashing proposal.

## Block Header

We would add two new fields in the block header:

```D
struct BlockHeader
{
    Hash prev_block;
    Hash height;
    Hash merkle_root;

    /// Hash of the merkle root of the preimages for this height
    Hash preimage_root;  // new

    /// List of indices to the validator UTXO set which have not revealed the preimage
    uint[] missing_preimage_validators; // new
}
```

- `missing_preimage_validators` (MPV for short) will contain the list of indices into the sorted array of enrolled UTXOs which belong to the validators which have **not** revealed their preimage.
- `preimage_root` will contain the merkleized hash of the sorted array of preimages, where the preimages are all the preimages for which the associated UTXO is **not** in the MPV array.

To validate the block header:
1. Validate the block signature as usual. The new `preimage_root` and `missing_preimage_validators` are also part of the hash, therefore they are not malleable.
2. Retrieve the array of preimages for all the enrolled UTXOs which are not in the MPV array. Merkleize this array to get the preimage root hash. If this root hash matches `preimage_root`, then the block header is valid. Otherwise, it's invalid.

If the node which is doing the catchup doesn't itself have the preimage for some given validators even if the `MPV` is empty, then it means either:
- The node will need to fetch the preimages via the `getPreimage (Hash enroll_key)` API endpoint.
- The block header is invalid, despite there being a valid signature for it. If there is `50% + 1` of validators in the `validators` bitmask in the block header then this should not be possible, otherwise the entire system is compromised.

### Why is MPV an array instead of a bitmask?

In the vast majority of cases we expect the number of validators which did not reveal preimages to be low. Take an example of a system with 1000 validators. If we were to use a bitmask, this would consume 125 bytes for each block header, despite most bits in this bitmask being 1 (or most being 0). If using an array instead, slashing one validator would only consume 5 bytes (1 byte VarInt for the length + 1 `uint` index). We can further optimize this depending on if we have an upper bound on the number of total validators in the system (for example `ushort[]` in place of `uint[]`).

## ConsensusData

We would add two new fields to the `ConsensusData`:

```D
public struct ConsensusData
{
    Transaction[] tx_set;
    Enrollment[] enrolls;

    /// Hash of the merkle root of the preimages for this height
    Hash preimage_root;  // new

    /// List of indices to the validator UTXO set which have not revealed the preimage
    uint[] missing_preimage_validators;  // new
}
```

The meaning of these is the same as in the `BlockHeader`. With one difference: they reflect the state of the node which has sent the `ConsensusData` message. So `ConsensusData` does not reflect the state of the entire system, not until the nodes have reached consensus on it through SCP.

### Nominating ConsensusData

A Validator creates and nominates a `ConsensusData` struct with the merkleized `preimage_root` of all the preimages it has for the block height which is about to be nominated. If the node is missing any preimages it will add their associated UTXO indices into the `MPV` inside the `ConsensusData`.

### Validating ConsensusData - `validateValue()`

When a Validator receives a `ConsensusData` struct it will try to verify the `preimage_root` is correct.

If there is a transaction in the `ConsensusData` which tries to melt a UTXO which is in the `MPV`, then this ConsensusData should not be accepted.

The validator uses the `MPV` array to construct an array of preimages and derive the merkleized root hash. 

1) If the validator is missing a preimage even though it's **not** in the received MPV it **might** mean that this validator simply hasn't received this preimage yet (e.g. networking outage). In this case AND if the transaction set is acceptable then the validator should return `kMaybeValidValue`. It is possible that the rest of the network does indeed have the preimage for a given validator, therefore the rest of the network may end up accepting this nomination in the end. If this node's quorum does indeed do that, then by SCP rules it will have to accept the nomination. This step is automatic and happens under the hood in SCP.
2) If the calculated preimage root matches the `preimage_root`, AND the transaction set is acceptable, then it has two choices: 
A) If the received MPV is larger than the validator's own MPV, it should return `kMaybeValidValue`. This means the validator itself has the missing preimage, but the other validator which proposed the consensus data does not. It's still possible that the rest of the network decides to accept the larger MPV proposal. So we cannot return `kInvalidValue` because that signals this proposal can never be accepted (e.g. bad signatures)
B) If the MPV is the same size as the validator's own MPV, and the transaction set is acceptable it should return `kValidValue`.
C) If the received MPV is smaller than the validator's own MPV, then this case should have already been handled by `1)` already.

## Tracking the preimage root & MPV

For efficiency, I recommend generating the `preimage_root` and `MPV` for the local validator node when it's time to start nominating a new block. Then, if new preimages are received during nomination then the node may update these values. The `prepareNominatingSet()` and `validateValue()` routines in the `Nominator` class should always use the latest `preimage_root` and `MPV`.

## Changing getRandomSeed

The [getRandomSeed](https://github.com/bpfkorea/agora/blob/260ba6305a6593e8a4bef528b33ac5ce6cc6986b/source/agora/consensus/EnrollmentManager.d#L536) function will be changed to simply return the `preimage_root` for the given block height.

## Slashing

Slashing is automatic and happens in the update routine of the UTXO set after an externalized block has been validated and is about to be accepted to the blockchain. For all the utxos in `MPV`, we modify the UTXO set and route the slashed funds to the commons budget.

There is still a question on how to keep the UTXO set valid after this operation. I'm not sure yet if this will be a problem or how to solve it.

# Benefits over the transaction-based slashing

- It eliminates complicating the transaction validation routines. Currently it's impossible to craft a transaction which "slashes" a UTXO because the node which generated the slashing transaction does not own the key-pair of the slashed UTXO, therefore signature validation fails. We would need to add special support for slashing transactions to make them valid.
- It avoids having many validators generate competing slashing transactions. Defining which node may craft the slashing transaction is complex.
- It adds a **proof** to the block header that the set of preimages have really existed for a block height when that block height was externalized, in the form of `preimage_root`. A slashing transaction-based approach could also have the `preimage_root` in the block header, but it would require SPVs to also download the slashing transactions themselves (and validate them).

**Comment by AndrejMitrovic** [comment]
By the way, the transaction-based slashing could in theory work, but we would need to add execution engine support for it. It would be a complex feature and the freeze transaction would have to change. It would have a lock script like follows:

```
IF
    <signature> <bitmask> OP.VERIFY_VALIDATOR_SIGNATURE
ELSE
    <signature> <key> OP.VERIFY_SIGNATURE
```

Where `OP.VERIFY_VALIDATOR_SIGNATURE` would accept only a bitmask that has >= `50% + 1` bits set, and where the signature would be valid if we used the sum of the public keys of the validators which are in the bitmask to validate it.

Then to construct the slashing transaction the unlock script would contain `OP.TRUE <bitmask> <signature>`. However it's not so easy. There is still the race condition where the `ELSE` branch could be accepted by a competing transaction. And the transaction's valid state in the engine would depend on the block height - so it's no longer possible to cache a transaction's script validation.

It might be an idea for the future, but requires a lot of thought..

**Comment by AndrejMitrovic** [comment]
```
If the calculated preimage root matches the preimage_root, AND the transaction set is acceptable, then it has two choices:
A) If the received MPV is larger than the validator's own MPV, it should return kMaybeValidValue. This means the validator itself has the missing preimage, but the other validator which proposed the consensus data does not. It's still possible that the rest of the network decides to accept the larger MPV proposal. So we cannot return kInvalidValue because that signals this proposal can never be accepted (e.g. bad signatures)
B) If the MPV is the same size as the validator's own MPV, and the transaction set is acceptable it should return kValidValue.
C) If the received MPV is smaller than the validator's own MPV, then this case should have already been handled by 1) already.
```

This section can be simplified. I suggest changing to as follows:

- Always keep the `mpv` arrays sorted, both locally and in the `ConsensusData` that the node creates (do not modify incoming `ConsensusData`).
- When validating `ConsensusData`, the `mpv` array in `ConsensusData` must be sorted. You may check that with `mpv.isStrictlyMonotonic()`.
- If the `mpv` in `ConsensusData` is exactly the same as ours and the transactions are valid, return `kValidValue`.
- If the arrays do not match, but the transactions are valid, return `kMaybeValidValue`.

By using sorting you can compare arrays easily with `this.mpv == cdata.mpv`.

Does that make sense?

**Comment by linked0** [comment]
It does! It makes the algorithm simple and neat.
Thanks.

**Comment by linked0** [comment]
After discussing with @AndrejMitrovic during code review of [the PR](https://github.com/bpfkorea/agora/pull/1412), there have been some changes in design.
1. We use a random seed of preimages of validators instead of a merkleized root of them.
2. We only add MPVs into the `ConsensusData`.
3. In the catch-up process, we need pre-images for un-enrolled validators inclusive of slashed ones. So we can leave the un-enrolled validator information in order to get the preimages needed to check the validity of the random seed in the `BlockHeader`

**Comment by AndrejMitrovic** [comment]
Closing this as most of the work has been done. Any leftover work will be split off into separate issues.
